I1170 combined gaussian log likelihood

DavAug opened the pull request:
Closes #1170. 

- [x] Implements CombinedGaussianLogLikelihood with a constant variance contribution and one that scales with model output.
- [x] Introduces 3 parameters per model output, `sigma_base`, `eta`, `sigma_rel`, see docstring for details
- [x] Implements call that agrees with `pints.GaussianLogLikelihood` and `pints.MultplicativeGaussianLogLikelihood` for appropriate parameter choices, see tests for details.
- [x] Implements evaluateS1 that agrees with `pints.GaussianLogLikelihood` for appropriate parameter choices, see tests for details. (Expressions are relatively complex but can be found in docstring).

MichaelClerx commented on pints/_log_likelihoods.py:
```diff
@@ -237,6 +237,240 @@ def __call__(self, x):
+
+    For a time series model :math:`f(t| \theta)` with parameters :math:`\theta`
+    , the CombinedGaussianLogLikelihood assumes that the model predictions
+    :math:`X` are Gaussian distributed
```
according to

DavAug commented:
@MichaelClerx 
Would you have an idea how I managed to break the timer?
```
======================================================================
FAIL: test_timing (test_timer.TestTimer)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/pints/pints/pints/tests/test_timer.py", line 39, in test_timing
    self.assertLess(b, a)
AssertionError: 1.7399999990175274e-05 not less than 9.900000009110954e-06
```

DavAug commented:
never mind it somehow didn't happen again.

MichaelClerx commented:
> @MichaelClerx
> Would you have an idea how I managed to break the timer?
> 
> ```
> ======================================================================
> FAIL: test_timing (test_timer.TestTimer)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/home/runner/work/pints/pints/pints/tests/test_timer.py", line 39, in test_timing
>     self.assertLess(b, a)
> AssertionError: 1.7399999990175274e-05 not less than 9.900000009110954e-06
> ```

Look @chon! Timers are weird!

chonlei commented:
Yup it's convinced me...! Weird!!

ben18785 commented on pints/_log_likelihoods.py:
```diff
@@ -237,6 +237,240 @@ def __call__(self, x):
         )
 
 
+class CombinedGaussianLogLikelihood(pints.ProblemLogLikelihood):
```
Just wondering about the name? Is this what this type of model is called in the literature or is it one we've made up? To me, it isn't that informative about the type of noise model (but that's perhaps because I'm not au fait with the literature). I would call it something like `MultiplicativeAndConstantGaussianLogLikelihood`.

DavAug replied:
In the PKPD community "combined Gaussian error" seems to be quite popular. I personally don't really mind the name too much, but `pints.MultiplicativeAndConstantGaussianLogLikelihood()` has already 52 characters. So simply from a flake8 perspective this name may be a little too long.

ben18785 commented on pints/_log_likelihoods.py:
```diff
@@ -237,6 +237,240 @@ def __call__(self, x):
+
+        # Compute log-likelihood
+        # (inner sums over time points, outer sum over parameters)
+        log_likelihood = self._logn - np.sum(
```
Looks good.

ben18785 commented on pints/_log_likelihoods.py:
```diff
@@ -237,6 +237,240 @@ def __call__(self, x):
+        parameters are
+
+        .. math::
+            \frac{\partial \log L}{\partial \theta _k}
```
I haven't been through these thoroughly, so would be good if someone else could check too. I am happy though that, as things currently stand, the tests show these derivs reproduce the special cases of the multiplicative model and GaussianLogLikelihood.

DavAug commented:
> @DavAug Looks good; thanks! My main comment was, can we add an example notebook? Apart from showing users how to use this functionality, I'm interested to know how easy it is to identify the parts of this noise model...
>
>Also, to others, it'd be good if someone could cast an eye of the derivatives too. They looked reasonable on first glances but I haven't been through with a fine tooth comb.

Thank you @ben18785 for the review. Where exactly would you like the notebook to be? We don't really have notebooks yet that are dedicating to LogLikelihoods do we? So I could add it to the toy directory, or to the the HMC and AMCMC notebook as another example, if we didn't want to add a new repo for this likelihood. Another option would be to open a new folder which we then could gradually fill with examples for all likelihoods and a fixed set of models and samplers.

MichaelClerx commented:
> > @DavAug Looks good; thanks! My main comment was, can we add an example notebook? Apart from showing users how to use this functionality, I'm interested to know how easy it is to identify the parts of this noise model...
> > Also, to others, it'd be good if someone could cast an eye of the derivatives too. They looked reasonable on first glances but I haven't been through with a fine tooth comb.
> 
> Thank you @ben18785 for the review. Where exactly would you like the notebook to be? We don't really have notebooks yet that are dedicating to LogLikelihoods do we? So I could add it to the toy directory, or to the the HMC and AMCMC notebook as another example, if we didn't want to add a new directory for this likelihood. Another option would be to open a new folder which we then could gradually fill with examples for all likelihoods and a fixed set of models and samplers.

There's a "Statistical modelling" section that has examples with different noise models (and the appropriate likelihoods)

DavAug commented:
> Thanks @DavAug. Think it looks really good apart from the HMC bit in the example notebook... Let me know when you've been able to compare these to finite difference estimates of gradients and will have another look.

>Can we also add the following to the various plots in the example notebook?

>Reference parameter values to the trace plots;
>Parameter names to the axes of the trace plots.

Thanks @ben18785. I added the finite difference tests, and corrected a bug in the partial derivative with respect to eta. The derivatives should now be correct for positive model outputs. Two things that are still not quite clear:
1. HMC still gives crappy results and I am not quite sure why. It may be that the step sizes are not chosen properly.
2. The partial derivative with respect to eta is only correct as long as the model output is positive (derivative involves a log(f) term). We could generalise by clarifying that the multplicative noise really scales with the magnitude of the model output, but this would introduce a discontinuity in the derivative of f. Would you rather restrict to positive output models or somehow try to deal with the disconinuities?

MichaelClerx commented:
Hey @DavAug ,

I'm looking at existing branches and deleting where appropriate. Was this merged via a different branch/PR? Or did we give up on this (in which case we can keep the branch for future ref)?

DavAug commented:
> Hey @DavAug ,
> 
> I'm looking at existing branches and deleting where appropriate. Was this merged via a different branch/PR? Or did we give up on this (in which case we can keep the branch for future ref)?

This has been merged, I am not entirely sure why it doesn't show up in the history.. Let me check

DavAug commented:
@MichaelClerx 
Ok, looks like I cloned the branch from a previous commit for some reason that I don't remember completely. Here is the branch that we merged into master #1190

DavAug commented:
I'm deleting this branch now :)
